content/blog/2025-12-16-stateful-allo.md

```diff
@@ -0,0 +1,115 @@
++++
+title = "Modeling Stateful Accelerators in Allo"
+[extra]
+bio = """
+  Sunwoo Kim is interested in AI hardware. He is pursuing ECE PhD at Cornell.
+"""
+[[extra.authors]]
+name = "Sunwoo Kim"
+link = "https://sunwookim028.github.io/"
++++
+
+# Context and Vision
+Computer architects are building accelerator chips for demanding domains like AI. 
+[Allo](https://cornell-zhang.github.io/allo/) is a Python-embedded domain-specific language to facilitate the design of such accelerators in a modular and composable manner, with users across industry and academia.
+However, the current language doesn't have a support for stateful variables, hence users cannot model practical accelerators that holds intermediate data or results on chip.
+We aimed to add a language feature for users to specify stateful variables with compilation flow to lower them to static variables in Vitis HLS backend.
+
+# Summary of Results
+We successfully added a new type qualifier in Allo to specify stateful variables and compilation flow to lower them into static variables in Vitis HLS, bridged by MLIR.
+We demonstrated programmable stateful accelerator designs for scalar and tensor arithmetics.
+Our most up-to-date code is in this [fork](https://github.com/sunwookim028/allo) and is being merged to the upstream.
+Here's a presentation [video](https://youtu.be/2dKNX0L-iG8?si=Jlyv5rDRoa-c0X9h) for a different course (ECE6775) where this project was also presented.
+
+# Acknowledgement
+This work was done with Joseph Maheshe, Jifeng Wu, Jenny Lee from ECE6775 project team.
+My specific contributions were in suggesting the frontend syntax, implementing the MLIR to HLS backend and writing and running stateful scalar and tensor arithmetic accelerators.
+
+# Design and Implementation
+<img src="./2025-12-16-stateful-allo/overview.png" alt="allo to mlir to hls" width="310"/>
+This is the overview of our compilation flow.
+We'll use this scalar accumulator example to explain each step.
+
+```python
+# A kernel that accumulates values across invocations
+def stateful_kernel(x: int32) -> int32:
+    # 'acc' retains its value between calls
+    acc: stateful(int32) = 0
+    acc = acc + x
+    return acc
+```
+In the frontend, adding the `stateful` qualifier specifies that the variable is a stateful one.
+
+```mlir
+module {
+  // Global storage (persistent)
+  memref.global "private" @__stateful_stateful_kernel_acc_1 : memref<i32> = dense<0>
+  func.func @stateful_kernel(%arg0: i32) -> i32 {
+    // 2. Local handle to storage
+    %0 = memref.get_global @__stateful_stateful_kernel_acc_1 : memref<i32>
+    // 3. Load/store operations using the handle
+    %1 = affine.load %0[] {from = "acc"} : memref<i32>
+    // ... arithmetic operations ...
+    affine.store %5, %0[] {to = "acc"} : memref<i32>
+    return %6 : i32
+  }
+}
+```
+They are then translated into global variables in IR, marked with a naming pattern `_stateful`.
```

Comment:
Why do they need to be marked with a special naming convention? Does it not suffice to have declared this variable with `memref.global`?